File: promptdev-backend/src/test/java/com/promptdev/service/JiraServiceTest.java
package com.promptdev.service;

import com.promptdev.dto.jira.*;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import org.springframework.http.MediaType;
import org.springframework.web.client.RestClient;

import java.util.List;
import java.util.Map;
import java.util.function.Function;

import static org.assertj.core.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
@SuppressWarnings({"unchecked", "rawtypes"})
class JiraServiceTest {

    @Mock
    private RestClient restClient;

    @Mock
    private RestClient.RequestHeadersUriSpec requestHeadersUriSpec;

    @Mock
    private RestClient.RequestHeadersSpec requestHeadersSpec;

    @Mock
    private RestClient.RequestBodyUriSpec requestBodyUriSpec;

    @Mock
    private RestClient.RequestBodySpec requestBodySpec;

    @Mock
    private RestClient.ResponseSpec responseSpec;

    private JiraService jiraService;

    @BeforeEach
    void setUp() {
        jiraService = new JiraService(restClient);
    }

    private void stubGetWithFunctionUri() {
        doReturn(requestHeadersUriSpec).when(restClient).get();
        doReturn(requestHeadersSpec).when(requestHeadersUriSpec).uri(any(Function.class));
        doReturn(responseSpec).when(requestHeadersSpec).retrieve();
    }

    private void stubGetWithTemplateUri() {
        doReturn(requestHeadersUriSpec).when(restClient).get();
        doReturn(requestHeadersSpec).when(requestHeadersUriSpec).uri(any(String.class), any(Object[].class));
        doReturn(responseSpec).when(requestHeadersSpec).retrieve();
    }

    private void stubPostChain(String uriTemplate) {
        doReturn(requestBodyUriSpec).when(restClient).post();
        doReturn(requestBodySpec).when(requestBodyUriSpec).uri(eq(uriTemplate), any(Object[].class));
        doReturn(requestBodySpec).when(requestBodySpec).contentType(MediaType.APPLICATION_JSON);
        lenient().doReturn(requestBodySpec).when(requestBodySpec).body(any());
        lenient().doReturn(requestBodySpec).when(requestBodySpec).body(any(Object.class));
        doReturn(responseSpec).when(requestBodySpec).retrieve();
        doReturn(null).when(responseSpec).toBodilessEntity();
    }

    private void stubPutChain(String uriTemplate) {
        doReturn(requestBodyUriSpec).when(restClient).put();
        doReturn(requestBodySpec).when(requestBodyUriSpec).uri(eq(uriTemplate), any(Object[].class));
        doReturn(requestBodySpec).when(requestBodySpec).contentType(MediaType.APPLICATION_JSON);
        lenient().doReturn(requestBodySpec).when(requestBodySpec).body(any());
        lenient().doReturn(requestBodySpec).when(requestBodySpec).body(any(Object.class));
        doReturn(responseSpec).when(requestBodySpec).retrieve();
        doReturn(null).when(responseSpec).toBodilessEntity();
    }

    @Nested
    @DisplayName("searchIssues")
    class SearchIssues {

        @Test
        @DisplayName("should call search endpoint with JQL, startAt, and maxResults")
        void shouldSearchWithJql() {
            stubGetWithFunctionUri();

            JiraSearchResponse expected = new JiraSearchResponse(0, 50, 1, List.of());
            doReturn(expected).when(responseSpec).body(JiraSearchResponse.class);

            JiraSearchResponse result = jiraService.searchIssues("project = TEST", 0, 50);

            assertThat(result).isNotNull();
            assertThat(result.total()).isEqualTo(1);
            assertThat(result.issues()).isEmpty();
            verify(restClient).get();
        }

        @Test
        @DisplayName("should return response with issues")
        void shouldReturnIssues() {
            stubGetWithFunctionUri();

            JiraIssueResponse issue = new JiraIssueResponse("10001", "TEST-1",
                    new JiraIssueResponse.JiraIssueFields(
                            "Test issue", null, null, null, null, null, null, null));
            JiraSearchResponse expected = new JiraSearchResponse(0, 50, 1, List.of(issue));
            doReturn(expected).when(responseSpec).body(JiraSearchResponse.class);

            JiraSearchResponse result = jiraService.searchIssues("project = TEST", 0, 50);

            assertThat(result.issues()).hasSize(1);
            assertThat(result.issues().get(0).key()).isEqualTo("TEST-1");
        }
    }

    @Nested
    @DisplayName("getIssue")
    class GetIssue {

        @Test
        @DisplayName("should fetch issue by key")
        void shouldFetchIssueByKey() {
            stubGetWithTemplateUri();

            JiraIssueResponse expected = new JiraIssueResponse("10042", "TEST-42",
                    new JiraIssueResponse.JiraIssueFields(
                            "Fix bug", null, null, null, null, "A bug description", null, null));
            doReturn(expected).when(responseSpec).body(JiraIssueResponse.class);

            JiraIssueResponse result = jiraService.getIssue("TEST-42");

            assertThat(result).isNotNull();
            assertThat(result.key()).isEqualTo("TEST-42");
            assertThat(result.fields().summary()).isEqualTo("Fix bug");
            assertThat(result.fields().description()).isEqualTo("A bug description");
            verify(restClient).get();
        }
    }

    @Nested
    @DisplayName("getTransitions")
    class GetTransitions {

        @Test
        @DisplayName("should return transitions for an issue")
        void shouldReturnTransitions() {
            stubGetWithTemplateUri();

            JiraTransitionResponse expected = new JiraTransitionResponse(List.of(
                    new JiraTransitionResponse.JiraTransition("11", "In Progress",
                            new JiraTransitionResponse.JiraStatus("In Progress", "3")),
                    new JiraTransitionResponse.JiraTransition("21", "Done",
                            new JiraTransitionResponse.JiraStatus("Done", "5"))
            ));
            doReturn(expected).when(responseSpec).body(JiraTransitionResponse.class);

            JiraTransitionResponse result = jiraService.getTransitions("TEST-10");

            assertThat(result).isNotNull();
            assertThat(result.transitions()).hasSize(2);
            assertThat(result.transitions().get(0).name()).isEqualTo("In Progress");
            assertThat(result.transitions().get(1).name()).isEqualTo("Done");
        }

        @Test
        @DisplayName("should return empty transitions list when none available")
        void shouldReturnEmptyTransitions() {
            stubGetWithTemplateUri();

            JiraTransitionResponse expected = new JiraTransitionResponse(List.of());
            doReturn(expected).when(responseSpec).body(JiraTransitionResponse.class);

            JiraTransitionResponse result = jiraService.getTransitions("TEST-99");

            assertThat(result.transitions()).isEmpty();
        }
    }

    @Nested
    @DisplayName("transitionIssue")
    class TransitionIssue {

        @Test
        @DisplayName("should POST transition to correct endpoint")
        void shouldPostTransition() {
            stubPostChain("/issue/{issueKey}/transitions");

            jiraService.transitionIssue("TEST-10", "21");

            verify(restClient).post();
            verify(requestBodySpec).body(any(JiraTransitionRequest.class));
            verify(responseSpec).toBodilessEntity();
        }
    }

    @Nested
    @DisplayName("addComment")
    class AddComment {

        @Test
        @DisplayName("should POST comment to correct endpoint")
        void shouldPostComment() {
            stubPostChain("/issue/{issueKey}/comment");

            jiraService.addComment("TEST-10", "This is a comment");

            verify(restClient).post();
            verify(requestBodySpec).body(any(JiraCommentRequest.class));
            verify(responseSpec).toBodilessEntity();
        }
    }

    @Nested
    @DisplayName("assignIssue")
    class AssignIssue {

        @Test
        @DisplayName("should PUT assignment to correct endpoint")
        void shouldPutAssignment() {
            stubPutChain("/issue/{issueKey}/assignee");

            jiraService.assignIssue("TEST-10", "jdoe");

            verify(restClient).put();
            verify(requestBodySpec).body(any(Map.class));
            verify(responseSpec).toBodilessEntity();
        }
    }

    @Nested
    @DisplayName("getIssuesByProject")
    class GetIssuesByProject {

        @Test
        @DisplayName("should search with correct project JQL")
        void shouldSearchByProject() {
            stubGetWithFunctionUri();

            JiraSearchResponse expected = new JiraSearchResponse(0, 10, 0, List.of());
            doReturn(expected).when(responseSpec).body(JiraSearchResponse.class);

            JiraSearchResponse result = jiraService.getIssuesByProject("MYPRJ", 0, 10);

            assertThat(result).isNotNull();
            verify(restClient).get();
        }
    }

    @Nested
    @DisplayName("getAssignedIssues")
    class GetAssignedIssues {

        @Test
        @DisplayName("should search with assignee JQL")
        void shouldSearchByAssignee() {
            stubGetWithFunctionUri();

            JiraSearchResponse expected = new JiraSearchResponse(0, 10, 0, List.of());
            doReturn(expected).when(responseSpec).body(JiraSearchResponse.class);

            JiraSearchResponse result = jiraService.getAssignedIssues("jdoe", 0, 10);

            assertThat(result).isNotNull();
            verify(restClient).get();
        }
    }
}

File: promptdev-ui/lib/services/__tests__/jira-service.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from "vitest";

const originalEnv = { ...process.env };

beforeEach(() => {
  process.env.JIRA_URL = "https://jira.example.com";
  process.env.JIRA_USERNAME = "admin";
  process.env.JIRA_TOKEN = "jira-token";
  process.env.JIRA_PROJECT_KEY = "PROJ";
  vi.clearAllMocks();
});

afterEach(() => {
  process.env = { ...originalEnv };
});

const mockFetch = vi.fn();
vi.stubGlobal("fetch", mockFetch);

import {
  isJiraConfigured,
  searchIssues,
  getIssue,
  getTransitions,
  transitionIssue,
  addComment,
  assignIssue,
  getIssuesByProject,
  getAssignedIssues,
} from "../jira-service";

function mockJsonResponse(data: unknown, status = 200) {
  return Promise.resolve({
    ok: status >= 200 && status < 300,
    status,
    text: () => Promise.resolve(JSON.stringify(data)),
  });
}

function mockEmptyResponse(status = 204) {
  return Promise.resolve({
    ok: true,
    status,
    text: () => Promise.resolve(""),
  });
}

describe("jira-service", () => {
  describe("isJiraConfigured", () => {
    it("should return true when JIRA_URL is set", () => {
      expect(isJiraConfigured()).toBe(true);
    });

    it("should return false when JIRA_URL is not set", () => {
      delete process.env.JIRA_URL;
      expect(isJiraConfigured()).toBe(false);
    });
  });

  describe("searchIssues", () => {
    it("should search with JQL and basic auth", async () => {
      const searchResult = {
        issues: [{ id: "1", key: "PROJ-1", fields: { summary: "Test" } }],
        startAt: 0,
        maxResults: 50,
        total: 1,
      };
      mockFetch.mockReturnValue(mockJsonResponse(searchResult));

      const result = await searchIssues("project = PROJ");

      expect(result).toEqual(searchResult);
      expect(mockFetch).toHaveBeenCalledWith(
        expect.stringContaining("/rest/api/2/search?jql="),
        expect.objectContaining({
          headers: expect.objectContaining({
            Authorization: expect.stringContaining("Basic "),
          }),
        }),
      );
    });

    it("should encode JQL in URL", async () => {
      mockFetch.mockReturnValue(mockJsonResponse({ issues: [], total: 0 }));

      await searchIssues("status = 'In Progress'");

      expect(mockFetch).toHaveBeenCalledWith(
        expect.stringContaining(encodeURIComponent("status = 'In Progress'")),
        expect.anything(),
      );
    });
  });

  describe("getIssue", () => {
    it("should fetch issue by key", async () => {
      const issue = { id: "1", key: "PROJ-1", fields: { summary: "Bug fix" } };
      mockFetch.mockReturnValue(mockJsonResponse(issue));

      const result = await getIssue("PROJ-1");

      expect(result).toEqual(issue);
      expect(mockFetch).toHaveBeenCalledWith(
        expect.stringContaining("/issue/PROJ-1"),
        expect.anything(),
      );
    });
  });

  describe("getTransitions", () => {
    it("should return transitions object", async () => {
      const transitions = {
        transitions: [
          {
            id: "21",
            name: "In Progress",
            to: { name: "In Progress", id: "3" },
          },
          { id: "31", name: "Done", to: { name: "Done", id: "5" } },
        ],
      };
      mockFetch.mockReturnValue(mockJsonResponse(transitions));

      const result = await getTransitions("PROJ-1");

      expect(result.transitions).toHaveLength(2);
      expect(result.transitions[0].name).toBe("In Progress");
    });
  });

  describe("transitionIssue", () => {
    it("should POST transition", async () => {
      mockFetch.mockReturnValue(mockEmptyResponse());

      await transitionIssue("PROJ-1", "21");

      expect(mockFetch).toHaveBeenCalledWith(
        expect.stringContaining("/issue/PROJ-1/transitions"),
        expect.objectContaining({
          method: "POST",
          body: JSON.stringify({ transition: { id: "21" } }),
        }),
      );
    });
  });

  describe("addComment", () => {
    it("should POST comment", async () => {
      mockFetch.mockReturnValue(mockEmptyResponse());

      await addComment("PROJ-1", "This is a comment");

      expect(mockFetch).toHaveBeenCalledWith(
        expect.stringContaining("/issue/PROJ-1/comment"),
        expect.objectContaining({
          method: "POST",
          body: JSON.stringify({ body: "This is a comment" }),
        }),
      );
    });
  });

  describe("assignIssue", () => {
    it("should PUT assignment", async () => {
      mockFetch.mockReturnValue(mockEmptyResponse());

      await assignIssue("PROJ-1", "developer");

      expect(mockFetch).toHaveBeenCalledWith(
        expect.stringContaining("/issue/PROJ-1/assignee"),
        expect.objectContaining({
          method: "PUT",
          body: JSON.stringify({ name: "developer" }),
        }),
      );
    });
  });

  describe("getIssuesByProject", () => {
    it("should search with project JQL", async () => {
      mockFetch.mockReturnValue(mockJsonResponse({ issues: [], total: 0 }));

      await getIssuesByProject("PROJ");

      expect(mockFetch).toHaveBeenCalledWith(
        expect.stringContaining(encodeURIComponent("project = PROJ")),
        expect.anything(),
      );
    });
  });

  describe("getAssignedIssues", () => {
    it("should search with assignee JQL", async () => {
      mockFetch.mockReturnValue(mockJsonResponse({ issues: [], total: 0 }));

      await getAssignedIssues("developer");

      expect(mockFetch).toHaveBeenCalledWith(
        expect.stringContaining(encodeURIComponent("assignee = developer")),
        expect.anything(),
      );
    });
  });

  describe("error handling", () => {
    it("should throw on HTTP error with response body", async () => {
      mockFetch.mockReturnValue(
        Promise.resolve({
          ok: false,
          status: 404,
          text: () => Promise.resolve("Issue not found"),
        }),
      );

      await expect(getIssue("PROJ-999")).rejects.toThrow(
        "Jira API error 404: Issue not found",
      );
    });

    it("should throw when JIRA_URL is not configured", async () => {
      delete process.env.JIRA_URL;
      await expect(searchIssues("test")).rejects.toThrow(
        "JIRA_URL is not configured",
      );
    });
  });
});
